src/Squidex.Infrastructure.MongoGridFs/Assets/MongoGridFsAssetStore.cs

```diff
@@ -0,0 +1,219 @@
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+using System.Threading.Tasks;
+using MongoDB.Driver;
+using MongoDB.Driver.GridFS;
+
+namespace Squidex.Infrastructure.Assets
+{
+    public class MongoGridFsAssetStore : IAssetStore, IInitializable
+    {
+        private readonly string path;
+        private readonly IGridFSBucket bucket;
+        private readonly DirectoryInfo directory;
+
+        public MongoGridFsAssetStore(IGridFSBucket bucket, string path)
+        {
+            Guard.NotNull(bucket, nameof(bucket));
+            Guard.NotNullOrEmpty(path, nameof(path));
+
+            this.bucket = bucket;
+            this.path = path;
+
+            directory = new DirectoryInfo(path);
+        }
+
+        public void Initialize()
+        {
+            try
+            {
+                // test bucket
+                bucket.Database.ListCollections();
+
+                if (!directory.Exists)
+                {
+                    directory.Create();
+                }
+            }
+            catch (MongoException ex)
+            {
+                throw new ConfigurationException(
+                    $"Cannot connect to Mongo GridFS bucket '${bucket.Options.BucketName}'.", ex);
+            }
+            catch (IOException ex)
+            {
+                if (!directory.Exists)
+                {
+                    throw new ConfigurationException($"Cannot access directory '{directory.FullName}'", ex);
+                }
+            }
+        }
+
+        public string GenerateSourceUrl(string id, long version, string suffix)
+        {
+            var file = GetFile(id, version, suffix);
+
+            return file.FullName;
+        }
+
+        public async Task CopyAsync(string name, string id, long version, string suffix,
+            CancellationToken ct = default(CancellationToken))
+        {
+            try
+            {
+                var file = GetFile(name);
+
+                using (var stream = new MemoryStream())
+                {
+                    await bucket.DownloadToStreamByNameAsync(file.Name, stream, cancellationToken: ct);
+                    await bucket.UploadFromStreamAsync(file.Name, stream, cancellationToken: ct);
+                }
+
+                file.CopyTo(GetPath(id, version, suffix));
+            }
+            catch (FileNotFoundException ex)
+            {
+                throw new AssetNotFoundException($"Asset {name} not found.", ex);
+            }
+            catch (GridFSException ex)
+            {
+                throw new AssetNotFoundException($"Asset {name} not found.", ex);
+            }
+        }
+
+        public async Task DownloadAsync(string id, long version, string suffix, Stream stream,
+            CancellationToken ct = default(CancellationToken))
+        {
+            var file = GetFile(id, version, suffix);
```

Comment:
As I said: if you want to improve performance it would make more sense to implement something like a CachingAssetStore that has an inner IAssetStore. Because you can use the caching for all other storages as well then. And if you want to cache it you have also to clean it up. Otherwise you could have a lot of files and you have no idea where all your disk space is gone.

Reply:
CachingAssetStore  -> agree

Reply:
where you want to cache the files? Memory?? You'll cache them on disk anyway, locally

Reply:
Depends how much cache levels you want.